graphql.go

```diff
@@ -73,12 +77,106 @@ func (c *Client) logf(format string, args ...interface{}) {
+	MaxInterval float64 `json:"maxInterval"`
+	// Optional - A mapping of statuses that client should retry.
+	// If not specifed, we will use default retry behavior on certain statuses
+	RetryStatus map[int]bool `json:"statusToRetry"`
+}
+
+// PolicyType defines a type of different possible Policies to be applied towards retrying
+type PolicyType string
+
+const (
+	// ExponentialBackoff - the interval is doubled after every try until hitting MaxInterval or MaxTries
+	ExponentialBackoff PolicyType = "exponential_backoff"
+	// Linear - the interval stays the same every try until hitting MaxTries
+	Linear PolicyType = "linear"
+)
+
+// Wrapper method to send request while optionally applying retry policy
+func (c *Client) sendRequest(req *http.Request) (*http.Response, error) {
+	var (
+		resp *http.Response
+		err  error
+	)
+
+	retryConfig := c.retryConfig
+	// Client did not specify retry config
+	if retryConfig.Policy == "" {
+		resp, err = c.httpClient.Do(req)
+		if err != nil {
+			return resp, fmt.Errorf("Error getting response: %v", err)
+		}
+		return resp, nil
+	}
+
+	for tryCount := 0; tryCount < retryConfig.MaxTries; tryCount++ {
+		resp, err = c.httpClient.Do(req)
+		if err == nil && !retryConfig.shouldRetry(resp.StatusCode) {
+			return resp, nil
+		}
+		log.Println("Will retry after interval expires")
+
+		// Wait for interval
+		log.Printf("Waiting for interval(%f) to expire...", retryConfig.Interval)
+		timer := time.NewTimer(time.Duration(retryConfig.Interval) * time.Second)
+		<-timer.C
+
+		// Increase interval
+		retryConfig.increaseInterval()
+		log.Printf("New interval: %f", retryConfig.Interval)
+	}
+
+	return nil, fmt.Errorf("Error getting response with retry: %s", err)
+}
+
+// Increase interval for exponential backoff policy until hitting MaxInterval
+func (config *RetryConfig) increaseInterval() {
+	if config.Policy == ExponentialBackoff && config.Interval < config.MaxInterval {
+		config.Interval = math.Min(config.Interval*2, config.MaxInterval)
+	}
+}
+
+// Determines whether the client should retry the request
+// If specified, the client will use consumer-specified RetryStatus to retry request based on status code
+// Otherwise, retry on 503, 504, and 507
+func (config *RetryConfig) shouldRetry(status int) bool {
+	if len(config.RetryStatus) > 0 {
+		return config.RetryStatus[status]
+	}
+	return status == http.StatusServiceUnavailable || status == http.StatusGatewayTimeout || status == http.StatusInsufficientStorage
+}
+
+// Determines whether RetryConfig is valid
+func (config *RetryConfig) isValid() bool {
+	isConfigOptional := config.Policy == ""
+	return isConfigOptional || (config.MaxTries > 0 && config.Interval <= config.MaxInterval)
+}
+
+// WithRetryConfig allows consumer to assign their retryConfig to the client's private retryConfig
+func WithRetryConfig(config RetryConfig) ClientOption {
+	return func(client *Client) {
+		client.retryConfig = config
+	}
+}
```

Comment:
we should have some sensible default values if the required ones aren't provided, or some default presets, like `DefaultLinearRetryPolicy` and `DefaultExponentialRetryPolicy` which already have the values set.

Reply:
I'll add that capability. Thanks!